### ISSUE TITLE
Overhaul of parrying

### ISSUE THREAD

**Issue by altf4** [altf4/SmashBot#58]
Now that SmashBot dashes, the implicit assumption behind the powershield logic is totally incorrect. It assumes that both characters remain stationary, while one attacks. This was never true, but was true enough in a walking Fox vs Marth world to be useful. But now that smashbot dashes, nothing there makes sense anymore. 

Furthermore, the parrying logic does not take into consideration at all attacks that have multiple hitboxes, or a single long one where the character moves. Lots of failures here.

Whatever new parrying paradigm we implement should be smart enough to know that given a character's current movement and state, where the hitboxes will be appearing in the future. (Modulo any unpredictable DI) This is not as simple as it sounds. How do you model a threat when an enemy Jigglypuff is doing a moving down-air at you? 

Lots to think about...

**Comment by geranim0** [comment]
The guy who did http://ikneedata.com/calculator has the functions to predict position in the air. Movement on the ground should be easy enough, a simple model could do the job. 

his reddit is https://www.reddit.com/user/schmooblidon
code https://github.com/schmooblidon/schmooblidon.github.io

**Comment by altf4** [comment]
That's good information, thanks! 

I don't see, however, character jumping trajectory data. I suspect I can reverse engineer that easily enough, but how exactly does the Melee physics engine handle trajectories of a character that's jumping in with an air attack?

**Comment by jalbatross** [comment]
Kind of an overlap with issue #55, but do you think it would be sufficient to be able to gather the hitbox position of every frame relative to Marth's position for an attack? I've been thinking about how to overhaul the parry system, and I thought that if we had that information then so long as the hitbox that we care about is within our shielding range we can powershield the attack and punish if needed and/or possible. 

For example,  we could run the following code which would run in place of the defend against attack sequence in Bait.cpp:
```
if (notShielding and notDamaged and marthIsAttacking) 
{
     if (inDanger (fox, marth) 
     {
           CreateTactic(parry);
     }
}

bool inDanger(defender, attacker)
{
     defenderX = defender.x;
     defenderY = defender.y;
     defenderFacing = defender.facing;
     //etc
     
    //it's assumed attacker is attacking
    //assume Marth here
     switch(attacker->actionState)
     {
       case (FAIR):
             //since no reverse hitbox for this move
             if (attackerNotFacingDefender) {break;}

            //if there is an active hitbox
            if (attacker.activeFrame <= lastHitFrame && attacker.activeFrame >= firstHitFrame)) 
           {
                 switch(attacker.activeFrame)
                 {
                       case firstHitFrame:
                           //in the case of Marth, the first hit of FAIR appears up and to the right
                           // of him. So the Fox's Y coordinate would have to be a little higher than
                          // Marth's since he's not very tall if we are going to be in its range
                           if (xDistance <= firstHitXRange && defenderY <= attackerY + firstHitYRange
                               && defenderY >= attackerY)  
                                 return true;
                      //for the second hit of Marth's fair the secondHitYRange value isn't too large because 
                     //is mostly horizontal. In this case it accounts for the 'height' of the hitbox
                      case secondHitFrame: 
                          if (xDistance <= secondHitXRange && defenderY <= attackerY + secondHitYRange
                             && defenderY >= attackerY)
                                    return true;
                         //etc.
                       
                 }
           }

     }

}


```

Obviously, this method has a number of disadvantages and it's extremely cumbersome, but I think it can work. So far I've only found that the maximum X-range of a Marth hitting a standing Fox's tail from behind with Fair is roughly 18.5 in-game X units. I only conducted this test with a falling Marth so there was still an active hitbox 'trailing' the Fair hitbox which would (ideally) also be accounted for if we approached the problem this way. Marth was within 1.4 Y-units of Fox when the attack landed.

As a (potentially) useful aside, I've discovered that Marth's standing height (the distance between his top-hitbox Y-coordinate and the y-coordinate of his static position marker) is approximately 20.9 in-game Y-units, while Fox's standing height is roughly 14.5 in-game Y-units. I think that these numbers could be useful for roughly gauging whether or not we can be hit by certain attacks (for example, we can escape from an attack if its hitbox is far enough above our head).

Let me know what you think! I'd love to help in any way that I can.

**Comment by altf4** [comment]
Hmm, that looks like it would work, but it more or less what we have currently. The existing powershield logic is good at knowing when to hit the L button given an incoming attack. The problem is that this isn't good enough. There are lots of situations where it's not safe to shield an attack, even if you do powershield it.

What we need is a system that can look into the future and recognize "There will be a hitbox at location X, with size Y, in exactly Z frames from now". SmashBot will then be able to calculate whether or not it's able to: A) Hit the enemy before it happens, B) Run away and put itself outside the hitbox area in time, C) Stay put and powershield / spotdodge / airdodge the attack. 

Doing so is easy for cases like Marth's forward smash. But it gets more complicated when you have to generalize that to consider all kinds of attacks. This logic should be generalizable for all attacks from any character.

**Comment by altf4** [comment]
In order to make this happen first, though, what we really need is a machine readable (probably CSV) file containing all of Melee's frame data. Comprehensively for all animations. 

All the stuff that is cumbersomely listed out inside GameState.cpp, and more. For each animation, we'd need max frames, IASA, first hitbox frame, last hitbox frame, size and relative position of each hitbox, frames of invincibility, whether or not you can be grabbed, etc...

Figuring out the spacing of every hitbox sounds extremely laborious and error prone. Perhaps there's a way to automate that process by looking at animated GIFs. Yea.... I like that idea. Could probably write a python script that looks for the red boxes and outputs a CSV with what frames they occur on and what positions they're in...

**Comment by altf4** [comment]
Closed for v2.0
